transports/quic/src/in_addr.rs

```diff
@@ -0,0 +1,121 @@
+use if_watch::{IfEvent, IfWatcher};
+
+use futures::{
+    future::{BoxFuture, FutureExt},
+    lock::Mutex,
+    stream::{Stream, StreamExt},
+};
+
+use std::{
+    io::Result,
+    net::IpAddr,
+    ops::DerefMut,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+/// Watches for interface changes.
+#[derive(Clone, Debug)]
+pub(crate) struct InAddr(Arc<Mutex<InAddrInner>>);
+
+impl InAddr {
+    /// If ip is specified then only one `IfEvent::Up` with IpNet(ip)/32 will be generated.
+    /// If ip is unspecified then `IfEvent::Up/Down` events will be generated for all interfaces.
+    pub(crate) fn new(ip: IpAddr) -> Self {
+        let inner = if ip.is_unspecified() {
+            let watcher = IfWatch::Pending(IfWatcher::new().boxed());
+            InAddrInner::Any {
+                if_watch: Box::new(watcher),
+            }
+        } else {
+            InAddrInner::One { ip: Some(ip) }
+        };
+        Self(Arc::new(Mutex::new(inner)))
+    }
+}
+
+/// The listening addresses of a `UdpSocket`.
+#[derive(Debug)]
+enum InAddrInner {
+    /// The socket accepts connections on a single interface.
+    One { ip: Option<IpAddr> },
+    /// The socket accepts connections on all interfaces.
+    Any { if_watch: Box<IfWatch> },
+}
+
+enum IfWatch {
+    Pending(BoxFuture<'static, std::io::Result<IfWatcher>>),
+    Ready(Box<IfWatcher>),
+}
```

Comment:
Referencing an discussion when `IfWatcher` was first added to tcp: https://github.com/libp2p/rust-libp2p/pull/1887#issuecomment-745233624.
`IfWatcher` used to (?) use `async_io` as reactor, which was not acceptable for some tokio users.  I believe this was now changed with https://github.com/mxinden/if-watch/commit/2b8dfcb9f645ebb8502d7db598a245f927213c0a so it should not be a problem anymore. But just to be sure, could you please confirm this assumption @mxinden @dvc94ch?

Reply:
Also: our `InAddr` logic for polling the `IfWatcher` is more or less a copy of the logic from the tcp transport (correct me if I am wrong @kpp). Admittedly it is not that much logic, but still duplicates some code. Even more if it turns out that we do need to have different implementations for `async_std` and `tokio` runtime. 
I think it would make sense to extract this to somewhere where both, the tcp transport and quic transport can use it. Either `libp2p-core`, or the `if-watch` crate directly (not familiar with the latter, so don't know if that makes sense). Wdyt @mxinden?

Reply:
> `IfWatcher` used to (?) use `async_io` as reactor, which was not acceptable for some tokio users. I believe this was changed with [mxinden/if-watch@2b8dfcb](https://github.com/mxinden/if-watch/commit/2b8dfcb9f645ebb8502d7db598a245f927213c0a) so it should not be a problem anymore.

As far as I understand this is correct. Though there is a discussion to make `if-watch` generic over the executor in-use: https://github.com/mxinden/if-watch/issues/21

Reply:
> I think it would make sense to extract this to somewhere where both, the tcp transport and quic transport can use it. Either `libp2p-core`, or the `if-watch` crate directly (not familiar with the latter, so don't know if that makes sense).

Moving this to `libp2p-core` makes sense to me. Off the top of my head, I think this is too libp2p specific in order to land in `if-watch`.

Reply:
> > I think it would make sense to extract this to somewhere where both, the tcp transport and quic transport can use it. Either `libp2p-core`, or the `if-watch` crate directly (not familiar with the latter, so don't know if that makes sense).
> 
> Moving this to `libp2p-core` makes sense to me. Off the top of my head, I think this is too libp2p specific in order to land in `if-watch`.

I think we should aim for `libp2p-core` to _not_ be a catch-all for common things. If something is only used by two transports, my first suggestion would be to create a separate crate that those two transports can depend on.

Reply:
Obsolete since with `ifwatcher-v2.0.0` the `IfWatcher` interface has simplified enough so that there is not much boilerplate code needed anymore.